GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs

```diff
@@ -1157,12 +1157,40 @@ private void UpdatePlaceholders()
                     //  1. Ensures child folders are deleted before their parents
                     //  2. Ensures that folders that have been deleted by git (but are still in the projection) are found before their
                     //     parent folder is re-expanded (only applies on platforms where EnumerationExpandsDirectories is true)
+                    Dictionary<string, long> availableSizes = new Dictionary<string, long>();
                     foreach (PlaceholderListDatabase.PlaceholderData folderPlaceholder in placeholderFoldersListCopy.OrderByDescending(x => x.Path))
                     {
-                        // Remove folder placeholders before re-expansion to ensure that projection changes that convert a folder to a file work
-                        // properly
-                        if (!this.RemoveFolderPlaceholderIfEmpty(folderPlaceholder, addPlaceholderToUpdatedPlaceholders, folderPlaceholdersToKeep))
+                        bool keepFolder = true;
+                        if (!folderPlaceholdersToKeep.Contains(folderPlaceholder.Path))
                         {
+                            bool isProjected = this.IsPathProjected(folderPlaceholder.Path, out string fileName, out bool isFolder);
+
+                            // Check the projection for the folder to determine if the folder needs to be deleted
+                            // The delete will be attempted if one of the following is true
+                            // 1. not in the projection anymore
+                            // 2. in the projection but is not a folder in the projection
+                            // 3. Folder no longer exist on disk - this is to remove the tombstone
+                            if (!isProjected ||
+                                !isFolder ||
+                                (!GVFSPlatform.Instance.KernelDriver.EnumerationExpandsDirectories &&

```

Comment:
Out of curiosity, why did the `EnumerationExpandsDirectories` check need to be added?

I would've thought that the `DirectoryExists` check wouldn't matter on on Mac.

Reply:
(I think it's the right change, I just wasn't sure why it would functionally break on Mac)

Reply:
I was trying to fix the `CheckoutBranchWithDirectoryNameSameAsFileWithWrite` test which is still failing in the tear down when validating the working directory.  Now there are 2 tests failing so this wasn't the right fix.  In the test it is checking out a branch with a file in file.txt/file.txt and edits that file then commits the change and checks out a branch with the file in the root at file.txt.  So it is the case of a folder changing to a file in the root directory and seems like the root directory is not re-expanding.

You wrote the expand on enumerate code so do you have any ideas what the issue might be?  Does it need to re-expand when folder changes to a file?  How and when was the root directory re-expanded in the previous code?

Reply:
> Does it need to re-expand when folder changes to a file?

Yes it will need to re-expand.

Previously I would have expected that it works like this:

1. git deletes file.txt/file.txt
2. VFS4G deletes file.txt (directory)
3. VFS4G re-expands root directory
4. VFS4G writes file.txt

But so far I've not found how your changes would have altered this behavior.

When I was debugging issues like this before (when implementing the re-expand logic).  I found it helpful to add a bunch of logging in this method so that VFS4G would record exactly what folders it's deleting (or not) and record what it does when re-expanding.

I'll spend some more time looking at your change and see if anything jumps out at me.

Reply:
> How and when was the root directory re-expanded in the previous code?

The root directory was expanded because it was in `placeholderFoldersListCopy`.  The expansion happened here:

`this.ReExpandFolder(blobSizesConnection, folderPlaceholder.Path, updatedPlaceholderDictionary, folderPlaceholders);`

On like 1202.

The root would have been expanded last as the folders are processed bottom up.